File: apps/ios/Wordocious/Sources/DailyCompletionsService.swift
import Foundation
import Supabase
import UIKit

/// One mode's daily result for today (for the completed-card state).
/// Codable so the store can cache today's completions on-device (web parity:
/// daily-completions-context.tsx seeds first render from sessionStorage so
/// completed badges never flash in after a fetch).
struct DailyCompletion: Codable {
    let gameMode: String
    let completed: Bool
    let guessCount: Int
    let timeSeconds: Double
    /// Per-mode daily composite score (daily_results.composite_score). Optional in
    /// the on-device cache written before this field existed → defaults to 0.
    let score: Double
    enum CodingKeys: String, CodingKey {
        case gameMode = "game_mode"
        case completed
        case guessCount = "guess_count"
        case timeSeconds = "time_seconds"
        case score = "composite_score"
    }
    init(gameMode: String, completed: Bool, guessCount: Int, timeSeconds: Double, score: Double = 0) {
        self.gameMode = gameMode; self.completed = completed
        self.guessCount = guessCount; self.timeSeconds = timeSeconds; self.score = score
    }
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        gameMode = try c.decode(String.self, forKey: .gameMode)
        completed = try c.decode(Bool.self, forKey: .completed)
        guessCount = try c.decode(Int.self, forKey: .guessCount)
        timeSeconds = try c.decode(Double.self, forKey: .timeSeconds)
        score = (try? c.decodeIfPresent(Double.self, forKey: .score)) ?? 0
    }
}

/// Summed totals across today's daily completions — one helper shared by the
/// banner, celebration modal, and share card so all three always agree.
struct DailyTotals {
    var completed: Int = 0
    var won: Int = 0
    let total: Int = DailyCompletionsStore.totalDailyModes
    var totalGuesses: Int = 0
    var totalTimeSeconds: Double = 0
    var totalScore: Double = 0
    var flawless: Bool { completed >= total && won >= total }

    init(_ byMode: [String: DailyCompletion]) {
        for c in byMode.values {
            completed += 1
            if c.completed { won += 1 }
            totalGuesses += c.guessCount
            totalTimeSeconds += c.timeSeconds
            // Web parity (daily-service.ts fetchTodayDailyCompletions): each
            // mode's score is rounded BEFORE summing — sum-of-rounds, not
            // round-of-sum, or the two platforms' sweep totals drift by ±1.
            totalScore += c.score.rounded()
        }
    }
}

/// Loads the signed-in player's own daily_results for today, keyed by mode.
/// Powers the home grid's completed states + the Flawless/Sweep banner.
/// Seeds the first render from an on-device cache (keyed by local day) so
/// cold launches don't flash unbadged cards while the network fetch runs.
@MainActor
final class DailyCompletionsStore: ObservableObject {
    @Published private(set) var byMode: [String: DailyCompletion] = [:]

    /// The 9 daily modes (VS excluded — no daily row).
    static let totalDailyModes = 9

    private static let cacheKey = "daily-completions-cache"

    var completedCount: Int { byMode.count }

    /// Today's cached completion count without spinning up a store — lets
    /// NotificationService decide whether tonight's reminder is still needed.
    /// The cache is day-keyed, so a stale (yesterday's) cache reads as 0.
    static func cachedTodayCount() -> Int { readCache()?.count ?? 0 }
    var wonCount: Int { byMode.values.filter { $0.completed }.count }
    var allDone: Bool { completedCount >= Self.totalDailyModes }
    var flawless: Bool { allDone && wonCount >= Self.totalDailyModes }

    /// Summed totals (points/time/guesses) across today's completions.
    var totals: DailyTotals { DailyTotals(byMode) }

    /// Posted by DailyResultsService the moment a daily finishes — the native
    /// analogue of the web's `daily-completion` window event, so the home grid
    /// flips to "completed" instantly instead of waiting for a refetch on the
    /// next tab switch.
    static let completionPosted = Notification.Name("wordocious.daily-completion")
    /// Posted AFTER the daily_results row is actually written to the server
    /// (completionPosted fires optimistically BEFORE the network call, so
    /// server-backed surfaces that refetch on it raced the insert and cached
    /// the pre-result leaderboard — the "rank doesn't show until I switch
    /// modes and back" bug).
    static let completionRecorded = Notification.Name("wordocious.daily-recorded")

    /// Finishes recorded during THIS session for the current local day (from the
    /// completionPosted note). Merged into a load() result to cover the
    /// read-after-write race where a just-INSERTed row isn't queryable yet.
    /// Day-scoped and cleared the moment the local day rolls over, so a finish
    /// from yesterday can never be carried into today — the bug that made a
    /// session left open across midnight show yesterday's sweep as today's.
    private var optimistic: [String: DailyCompletion] = [:]
    private var optimisticDay: String = LeaderboardService.todayLocal()

    init() {
        byMode = Self.readCache() ?? [:]
        NotificationCenter.default.addObserver(forName: Self.completionPosted, object: nil, queue: .main) { [weak self] note in
            guard let self, let c = note.object as? DailyCompletion else { return }
            // Best-result semantics: never downgrade a recorded win on replay.
            if let existing = self.byMode[c.gameMode], existing.completed && !c.completed { return }
            self.byMode[c.gameMode] = c
            self.optimistic[c.gameMode] = c
            self.optimisticDay = LeaderboardService.todayLocal()
            Self.writeCache(self.byMode)
            WidgetBridge.update(completions: self.byMode)
        }
        // A session left open across LOCAL midnight must refresh to the new day's
        // (empty) completions. Without this, yesterday's byMode lingered and — via
        // the old in-memory merge — got re-stamped onto today, so the home grid
        // showed yesterday's sweep as done with no board behind it. Reload on
        // foreground + on the system day change, mirroring web daily-boundary-reload.
        let reload: (Notification) -> Void = { [weak self] _ in Task { await self?.load() } }
        NotificationCenter.default.addObserver(forName: UIApplication.willEnterForegroundNotification, object: nil, queue: .main, using: reload)
        NotificationCenter.default.addObserver(forName: .NSCalendarDayChanged, object: nil, queue: .main, using: reload)
    }

    func load() async {
        let today = LeaderboardService.todayLocal()
        // Day rolled over since our last optimistic finish → those finishes are
        // yesterday's; drop them before they can be re-merged onto today.
        if optimisticDay != today { optimistic = [:]; optimisticDay = today }

        let client = AuthService.shared.client
        guard (try? await client.auth.session) != nil,
              let userId = try? await client.auth.session.user.id.uuidString else {
            byMode = [:]; optimistic = [:]; Self.writeCache(nil); return
        }
        do {
            let rows: [DailyCompletion] = try await client.from("daily_results")
                .select("game_mode, completed, guess_count, time_seconds, composite_score")
                .eq("user_id", value: userId)
                .eq("day", value: today)
                .eq("play_type", value: "solo")
                .execute().value
            // The SERVER is authoritative for today. Re-add ONLY this-session
            // optimistic finishes the server hasn't surfaced yet (the read-after-
            // write race: `.onDailyCompletion` calls load() the instant a daily
            // finishes, before the INSERT is queryable — a blind replace would drop
            // it and show a real Flawless as N-1/9). Crucially we NO LONGER merge
            // stale in-memory/cached state, which is how yesterday's completions
            // leaked into today. A day rollover clears `optimistic`, so a fetch
            // that returns nothing correctly yields an empty (fresh) board.
            var merged = Dictionary(rows.map { ($0.gameMode, $0) }, uniquingKeysWith: { a, _ in a })
            for (k, v) in optimistic where merged[k] == nil { merged[k] = v }
            byMode = merged
            Self.writeCache(byMode)
            WidgetBridge.update(completions: byMode)
        } catch {
            // Keep the cached state on a transient failure instead of blanking.
        }
    }

    // MARK: Day-keyed cache (the iOS analogue of the web's sessionStorage seed)

    private struct Cache: Codable { let day: String; let byMode: [String: DailyCompletion] }

    private static func readCache() -> [String: DailyCompletion]? {
        guard let data = UserDefaults.standard.data(forKey: cacheKey),
              let cache = try? JSONDecoder().decode(Cache.self, from: data),
              cache.day == LeaderboardService.todayLocal() else { return nil }
        return cache.byMode
    }

    private static func writeCache(_ byMode: [String: DailyCompletion]?) {
        guard let byMode else { UserDefaults.standard.removeObject(forKey: cacheKey); return }
        if let data = try? JSONEncoder().encode(Cache(day: LeaderboardService.todayLocal(), byMode: byMode)) {
            UserDefaults.standard.set(data, forKey: cacheKey)
        }
    }
}

// formatShortTime now comes from WordociousCore (Format.swift) — the local
// copy rendered "—" at 0s where web/Android render "0s".

/// Seconds until the next LOCAL midnight (puzzles reset locally).
func secondsUntilLocalMidnight() -> Int {
    let cal = Calendar.current
    guard let next = cal.nextDate(after: Date(), matching: DateComponents(hour: 0, minute: 0, second: 0), matchingPolicy: .nextTime) else {
        return 0
    }
    return max(0, Int(next.timeIntervalSinceNow))
}

import SwiftUI

extension View {
    /// Re-run `action` the instant a daily game is recorded (the
    /// `DailyCompletionsStore.completionPosted` notification), so completed-state
    /// surfaces — leaderboard, profile, records, home — refresh immediately
    /// instead of only on the next tab switch / re-navigation. The notification
    /// fires even when the view is in a backgrounded tab (its body stays alive),
    /// so by the time the user navigates over, the data is already current.
    func onDailyCompletion(_ action: @escaping () -> Void) -> some View {
        onReceive(NotificationCenter.default.publisher(for: DailyCompletionsStore.completionPosted)) { _ in action() }
    }

    /// Re-run `action` once the daily result row has LANDED on the server —
    /// the right trigger for surfaces that refetch server data (leaderboard
    /// rows/rank, records, completed-daily card). Listening to the optimistic
    /// completionPosted instead re-fetched BEFORE the insert and cached stale
    /// pre-result data.
    func onDailyRecorded(_ action: @escaping () -> Void) -> some View {
        onReceive(NotificationCenter.default.publisher(for: DailyCompletionsStore.completionRecorded)) { _ in action() }
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/DailyCompletionsService.kt
package com.wordocious.app.data

import com.wordocious.app.todayLocalDate
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Fetches the signed-in user's daily completions for today — mirrors the web
 * DailyCompletionsProvider (apps/web/lib/daily-completions-context.tsx). Drives
 * the Home mode-card W/L badges + completed-tint and the daily-sweep banner.
 */
object DailyCompletionsService {
    private val client get() = SupabaseConfig.client
    private val prefs by lazy {
        com.wordocious.app.App.instance.getSharedPreferences("wordocious_prefs", android.content.Context.MODE_PRIVATE)
    }
    private const val CACHE_KEY = "daily-completions-cache"
    private const val CACHE_DAY_KEY = "daily-completions-cache-day"

    /** Bumped the instant a daily is recorded (noteCompletion) — the Android
     *  analogue of the iOS completionPosted notification. Completed-state screens
     *  (home, leaderboard, profile, records) key their fetches on this so a
     *  finished puzzle shows immediately, without a tab round-trip. */
    private val _completionTick = kotlinx.coroutines.flow.MutableStateFlow(0)
    val completionTick: kotlinx.coroutines.flow.StateFlow<Int> = _completionTick

    @Serializable
    data class Completion(
        @SerialName("game_mode") val gameMode: String,
        val completed: Boolean,
        @SerialName("guess_count") val guessCount: Int = 0,
        @SerialName("time_seconds") val timeSeconds: Int = 0,
        /** Per-mode daily composite score (daily_results.composite_score). */
        @SerialName("composite_score") val score: Double = 0.0,
    )

    /**
     * Summed totals across today's completions — one helper shared by the
     * banner, celebration modal, and share card so all three always agree.
     */
    data class Totals(
        val completed: Int,
        val won: Int,
        val total: Int,
        val totalGuesses: Int,
        val totalTimeSeconds: Int,
        val totalScore: Int,
    ) {
        val flawless: Boolean get() = completed >= total && won >= total
    }

    /** The 9 daily modes (VS excluded — no daily row). */
    const val TOTAL_DAILY_MODES = 9

    fun totals(byMode: Map<String, Completion>): Totals {
        var won = 0; var guesses = 0; var time = 0; var score = 0.0
        for (c in byMode.values) {
            if (c.completed) won++
            guesses += c.guessCount; time += c.timeSeconds; score += c.score
        }
        return Totals(byMode.size, won, TOTAL_DAILY_MODES, guesses, time, Math.round(score).toInt())
    }

    /** The local day the last fetch served — lets [refreshIfDayChanged] detect a
     *  midnight rollover on a home screen left composed overnight. */
    private var lastServedDay: String? = null

    /** Bump the tick if the local day rolled over since the last fetch, so a home
     *  screen kept composed across midnight refetches the new day's (empty)
     *  completions instead of showing yesterday's. Call from an ON_RESUME hook. */
    fun refreshIfDayChanged() {
        val today = todayLocalDate()
        if (lastServedDay != null && lastServedDay != today) _completionTick.value++
        lastServedDay = today
    }

    /** Map of game_mode → completion for today's daily (solo). Empty if not signed in. */
    suspend fun fetchTodayCompletions(): Map<String, Completion> {
        lastServedDay = todayLocalDate()
        val userId = AuthService.userId ?: return emptyMap()
        return runCatching {
            val map = client.postgrest["daily_results"]
                .select(Columns.raw("game_mode,completed,guess_count,time_seconds,composite_score")) {
                    filter {
                        eq("user_id", userId)
                        eq("day", todayLocalDate())
                        eq("play_type", "solo")
                    }
                }
                .decodeList<Completion>()
                .associateBy { it.gameMode }
            // Read-after-write race: a fetch fired the instant a daily finishes
            // (completionTick) may not see the row we just INSERTed — a blind
            // replace would drop the just-finished mode, making a real Flawless
            // show as an N-1/9 Sweep. Keep any optimistic noteCompletion entry
            // (in the day-keyed cache) the server response is still missing.
            val merged = map.toMutableMap()
            for ((k, v) in readCache()) if (k !in merged) merged[k] = v
            writeCache(merged)
            // G6: fire-and-forget AFTER the completions map is ready (never
            // delays the return/publication) — warm the recorded-match row for
            // every daily already played today, so opening one on this device
            // replays instantly instead of flashing an empty board while
            // GameScreen's network fetch runs. Idempotent per seed per session.
            prefetchScope.launch {
                for (modeName in merged.keys) {
                    GameResultsService.prefetchRecordedDailyMatch(com.wordocious.app.todayLocalSeed(modeName))
                }
            }
            merged
        }.getOrElse { readCache() }   // transient failure → keep cached state
    }

    private val prefetchScope = kotlinx.coroutines.CoroutineScope(
        kotlinx.coroutines.SupervisorJob() + kotlinx.coroutines.Dispatchers.IO
    )

    // ── Day-keyed cache (web sessionStorage parity) — seeds the home grid's
    // first render so cold launches don't flash unbadged cards while the
    // network fetch runs.

    /**
     * Optimistic local update the moment a daily finishes (web 'daily-completion'
     * event parity): writes the completion into the day-keyed cache so the home
     * grid shows "completed" instantly when the player returns — HomeScreen
     * seeds from this cache on every (re)composition. Never downgrades a win.
     */
    fun noteCompletion(gameMode: String, completed: Boolean, guessCount: Int, timeSeconds: Int, score: Double = 0.0) {
        val current = readCache().toMutableMap()
        val existing = current[gameMode]
        if (existing != null && existing.completed && !completed) return
        current[gameMode] = Completion(gameMode = gameMode, completed = completed, guessCount = guessCount, timeSeconds = timeSeconds, score = score)
        writeCache(current)
        _completionTick.value++
    }

    fun readCache(): Map<String, Completion> {
        if (prefs.getString(CACHE_DAY_KEY, null) != todayLocalDate()) return emptyMap()
        val json = prefs.getString(CACHE_KEY, null) ?: return emptyMap()
        return runCatching {
            kotlinx.serialization.json.Json.decodeFromString<List<Completion>>(json).associateBy { it.gameMode }
        }.getOrElse { emptyMap() }
    }

    /** Drop the cached completions — called on sign-out so a guest (or a
     *  different account) never seeds the home grid from the prior user's data. */
    fun clearCache() {
        runCatching { prefs.edit().remove(CACHE_KEY).remove(CACHE_DAY_KEY).apply() }
        // Prefetched replay rows are per-user too — never leak across accounts.
        GameResultsService.clearPrefetchedDailyMatches()
    }

    private fun writeCache(map: Map<String, Completion>) {
        runCatching {
            prefs.edit()
                .putString(CACHE_KEY, kotlinx.serialization.json.Json.encodeToString(
                    kotlinx.serialization.builtins.ListSerializer(Completion.serializer()), map.values.toList()))
                .putString(CACHE_DAY_KEY, todayLocalDate())
                .apply()
        }
    }
}
